% Generated by roxygen2 (4.1.1): do not edit by hand
% Please edit documentation in R/seurat.R
\name{feature.plot}
\alias{feature.plot}
\title{Visualize 'features' on a dimensional reduction plot}
\usage{
feature.plot(object, features.plot, dim.1 = 1, dim.2 = 2,
  cells.use = NULL, pt.size = 1, cols.use = heat.colors(10),
  pch.use = 16, reduction.use = "tsne", nCol = NULL)
}
\arguments{
\item{object}{Seurat object}

\item{features.plot}{Vector of features to plot}

\item{dim.1}{Dimension for x-axis (default 1)}

\item{dim.2}{Dimension for y-axis (default 2)}

\item{cells.use}{Vector of cells to plot (default is all cells)}

\item{pt.size}{Adjust point size for plotting}

\item{cols.use}{Ordered vector of colors to use for plotting. Default is
heat.colors(10).}

\item{pch.use}{Pch for plotting}

\item{reduction.use}{Which dimensionality reduction to use. Default is
"tsne", can also be "pca", or "ica", assuming these are precomputed.}

\item{nCol}{Number of columns to use when plotting multiple features.}
}
\value{
No return value, only a graphical output
}
\description{
Colors single cells on a dimensional reduction plot according to a 'feature'
(i.e. gene expression, PC scores, number of genes detected, etc.)
}
\details{
To determine the color, the feature values across all cells are placed into
discrete bins, and then assigned a color based on cols.use. The number of
bins is determined by the number of colors in cols.use
}

